Staging sites: Show a confirmation form before synchronizing

epeicher opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Resolves DOTCOM-13932
Resolves DOTCOM-13962
Resolves DOTCOM-13961

## Proposed Changes

* Adds a warning when the site to be synced is a Woo site and the database is checked
* Adds a domain confirmation input when the database is checked
* Unselects by default the values to be synced so they needs to be explicitly checked

|Before | After|
|-------|------|
| <img width="1678" height="1362" alt="CleanShot 2025-07-23 at 16 44 30@2x" src="https://github.com/user-attachments/assets/98c45194-a8b1-4c00-a126-2fcca583ea8f" />|  <img width="1866" height="1564" alt="CleanShot 2025-07-24 at 18 58 22@2x" src="https://github.com/user-attachments/assets/c4e30245-c794-4661-9ba2-93255c9c2d02" />|

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* To notify the user about dangerous changes

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
#### WooCommerce site
* Apply these changes and start Calypso locally
* Open a site with WooCommerce installed and active, e.g. `/overview/:siteSlug`
* Add a Staging site to it by clicking `Add Staging site +`
* Once the Staging site exists, select `Sync` dropdown on the top-right of the page
* Select `Pull from Staging`
* Check that there are no nodes selected and that the `Pull` button is disabled
* Select the database node
* Check that the Warning and Domain confirmation input are displayed
* Check the `Pull` button is disabled
* Insert the proper domain and check that the `Pull` button activates. Try with valid/invalid values.
* Unselect the database and select `Files and folders`
* Check the Pull button is enabled
* Check different combinations of selecting/unselecting Database and `Files and folders`. Also, expand `All files and folders` and select individual nodes
* Check that the Warnings and Domain confirmation input appears/disappears as expected, as well as the `Pull` button is enabled/disabled as expected.
- Close the modal and on the `Sync` dropdown select Push to Staging
- Check `Database tables`  
- Check that the Domain confirmation input is displayed and the `Push` button is disabled
- Check that the Domain confirmation input appears/disappears as expected, as well as the `Push` button is enabled/disabled as expected.
* Open the staging site and repeat the above steps
* This time, the WooCommerce warning should be displayed when selecting `Push to production`


#### Non WooCommerce site
- Repeat previous steps and check that the WooCommerce warning is not displayed
- Check that the Domain confirmation input is displayed as expected as well as the `Pull` button activation

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

epeicher committed:
Staging Sync: Add warnings for Woo stores

epeicher committed:
Staging Sync: Add domain confirmation input

epeicher committed:
Fix linting errors and rename variables

epeicher commented on client/sites/staging-site/components/staging-site-sync-modal/index.tsx:
```diff
@@ -314,6 +311,11 @@ export default function SyncModal( {
 		[ dispatch, querySiteId ]
 	);
 
+	const handleDomainConfirmation = useCallback(
```
Extracted to this small function because using `setDomainConfirmation` directly in the `onChange` [in line 449](https://github.com/Automattic/wp-calypso/pull/104857/files#diff-f35b2f1fb4fd4096dc82772ce4d3aa14a569548a3f12fdf7325090d8403b2432R449) was returning linting errors as it accepts the value as `undefined`

sejas commented on client/sites/staging-site/components/staging-site-sync-modal/index.tsx:
```diff
@@ -213,11 +215,13 @@ export default function SyncModal( {
-			// If nodes don't exist yet, default to 'checked' since we set the root to checked by default
-			return 'checked';
+			// If nodes don't exist yet, default to 'unchecked' since we set the root to unchecked by default
+			return 'unchecked';
```
I like this behavior too. We can use the same approach in Studio 👍 .

gcsecsey replied:
I agree, this is better UX IMO.

sejas commented on client/sites/staging-site/components/staging-site-sync-modal/index.tsx:
```diff
@@ -410,10 +421,35 @@ export default function SyncModal( {
+								) }
+							</Notice>
+						) }
+						{ showDomainConfirmation && (
+							<VStack>
+								<HStack alignment="left" spacing={ 1 }>
+									<Text>{ __( "Enter your site's name" ) }</Text>
+									<Text color="#D63638">{ productionSiteSlug }</Text>
+									<Text>{ __( 'to confirm.' ) }</Text>
+								</HStack>
+								<InputControl onChange={ handleDomainConfirmation } />
+							</VStack>
+						) }
```
I think we can use the label prop from InputControl directly.
I also observed a warning in the console about we should use `__next40pxDefaultSize` prop.

```suggestion
						{ showDomainConfirmation && (
							<InputControl
								className="staging-site-card__domain-confirmation-input"
								label={
									<Text>
										{ createInterpolateElement(
											__( "Enter your site's name <SiteName/> to confirm." ),
											{
												SiteName: <Text color="var(--studio-red-50)">{ productionSiteSlug }</Text>,
											}
										) }
									</Text>
								}
								onChange={ handleDomainConfirmation }
								__next40pxDefaultSize
							/>
						) }
```

And then the CSS would be:

```css
.staging-site-card__domain-confirmation-input label span {
	text-transform: none;
}
```

epeicher replied:
Thanks for the suggestion! That was my first attempt, but I changed back to an external label to avoid adding additional CSS. Do you think that is a better approach? (Update: Yes, for accessibility, I missed that in your comment above 😄 )

Regarding the warning, that's a great point, thanks for that! I'm adding that `__next40pxDefaultSize` prop

epeicher replied:
I have decided to apply the suggestion and use `style` as it seems to be the guideline for the new dashboard, please let me know if you disagree 😄 Change here [ef56a9b](https://github.com/Automattic/wp-calypso/pull/104857/commits/ef56a9bd6fa9ba23aa18632d83b4f17f5e66ee7f)

sejas reviewed (approved):
It works great, thanks for improving the pull/push dialog. I confirm I see:
- The checkboxes disabled by default
- The action button (pull, push) disabled when everything is disabled. 
- The text input to confirm the action where I need to write my domain if I want to sync the database.

I think you mentioned the same issue twice in the description. Probably you wanted to close DOTCOM-13961.

I left a suggestion about improving accessibility in the TextControl input.


https://github.com/user-attachments/assets/60f0eb8e-fb64-409b-afca-e4c2110177af

sejas commented on client/sites/staging-site/components/staging-site-sync-modal/index.tsx:
```diff
@@ -410,10 +421,35 @@ export default function SyncModal( {
+							<VStack>
+								<HStack alignment="left" spacing={ 1 }>
+									<Text>{ __( "Enter your site's name" ) }</Text>
+									<Text color="#D63638">{ productionSiteSlug }</Text>
```
Instead of hardcoding the hex color, we can use the color `var(--studio-red-50)`.

epeicher replied:
Thanks for the suggestion! Adding it

epeicher replied:
Added as part of [ef56a9b](https://github.com/Automattic/wp-calypso/pull/104857/commits/ef56a9bd6fa9ba23aa18632d83b4f17f5e66ee7f)

gcsecsey reviewed (approved):
Thanks for tackling this @epeicher, this is working really well. As far as I could test, it displays the WooCommerce warning and Domain confirmation and also disables the Pull/Push button correctly in each case.

https://github.com/user-attachments/assets/65c271f1-f8ab-43aa-a86a-1bf1f83234bc

WooCommerce site pushing to Staging:

<img width="1336" height="1256" alt="Image" src="https://github.com/user-attachments/assets/c073b1a2-daf7-4e2c-aa97-f0f0952d02a9" />

Non WooCoomerce site pushing to Production:

<img width="1336" height="1110" alt="Image" src="https://github.com/user-attachments/assets/25b6d635-002e-44f3-8ec6-4ce2811caf24" />

wojtekn commented:
@epeicher looks good. Can we use the proper warning?

<img width="833" height="178" alt="Screenshot 2025-07-24 at 08 46 39" src="https://github.com/user-attachments/assets/f96fc960-dd09-4239-9b64-d68baa3f144e" />

We could also add 'Learn more' pointing to https://developer.wordpress.com/docs/developer-tools/staging-sites/sync-staging-production/#staging-to-production at the end of the section. I know this warning is a bit lengthy, but we can tune it before launch.

epeicher committed:
Apply color value from library. Use label prop in Input

epeicher committed:
Update warning message and change Learn more URL

epeicher commented:
> @epeicher looks good. Can we use the proper warning?
> 
> <img alt="Screenshot 2025-07-24 at 08 46 39" width="833" height="178" src="https://private-user-images.githubusercontent.com/727413/470150683-f96fc960-dd09-4239-9b64-d68baa3f144e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMzNTI0MzgsIm5iZiI6MTc1MzM1MjEzOCwicGF0aCI6Ii83Mjc0MTMvNDcwMTUwNjgzLWY5NmZjOTYwLWRkMDktNDIzOS05YjY0LWQ2OGJhYTNmMTQ0ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyNFQxMDE1MzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMTJlODQzMmFmZjRlZjVjYTIzOTNmODIyYmQ1Yjg2N2M0YjIzMTJkZDFlYzMyYTBlN2MyZWZmYTZjYTAxMzUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.QYgsMjYmqtHqwstsqmTXXDcoXnZtNbYR2lQOmn6S4DE">
> We could also add 'Learn more' pointing to https://developer.wordpress.com/docs/developer-tools/staging-sites/sync-staging-production/#staging-to-production at the end of the section. I know this warning is a bit lengthy, but we can tune it before launch.

Thanks @wojtekn, I have added the proper warning now. I have also added the `Learn more` link pointing to the URL you suggest. I have tried to use `localizeURL` but it does not work properly for https://developer.wordpress.com/ as it points to https://developer.wordpress.com/es/docs/developer-tools/staging-sites/sync-staging-production/#staging-to-production instead of  https://developer.wordpress.com/es/docs/herramientas-para-desarrolladores/sitios-de-pruebas/#sincronizacion-de-pruebas-a-produccion

If you agree with those changes and with the other changes I can progress with the merge, please let me know

epeicher committed:
Add warning header

epeicher committed:
Fix type error in Text component

epeicher committed:
Hide the confirmation when there are no files or database selected

epeicher commented:
Thanks for the review @wojtekn! I have created https://github.com/Automattic/wp-calypso/pull/104949 to address that
